core/src/main/java/io/questdb/griffin/engine/functions/GroupByFunction.java

```diff
@@ -37,9 +37,29 @@ public interface GroupByFunction extends Function, Mutable {
     default void clear() {
     }
 
-    void computeFirst(MapValue mapValue, Record record);
+    /**
+     * Performs the first aggregation within a group.
+     * <p>
+     * Row id is provided for aggregation functions that consider row order, such as first/last.
+     * {@link Record#getRowId()} shouldn't be used for this purpose since not all records implement it.
+     *
+     * @param mapValue map value holding the group
+     * @param record   record holding the aggregated row
+     * @param rowId    row id, guaranteed to be growing; the value may be different from record.getRowId()
+     */
+    void computeFirst(MapValue mapValue, Record record, long rowId);
 
-    void computeNext(MapValue mapValue, Record record);
+    /**
+     * Performs a subsequent aggregation within a group.
+     * <p>
+     * Row id is provided for aggregation functions that consider row order, such as first/last.
+     * {@link Record#getRowId()} shouldn't be used for this purpose since not all records implement it.
+     *
+     * @param mapValue map value holding the group
+     * @param record   record holding the aggregated row
+     * @param rowId    row id, guaranteed to be growing; the value may be different from record.getRowId()
```

Comment:
nit: could you describe what rowId could be used for. E.g. what's the contract on storing/retrieving. It is a little opaque from the existing description.

Reply:
There is the following text in this javadoc:
> Row id is provided for aggregation functions that consider row order, such as first/last.
> {@link Record#getRowId()} shouldn't be used for this purpose since not all records implement it.

Do you think it needs to be expanded?

Reply:
i think so, it is unclear how this row id can be used in a sharded scenario. Doc should emphasise that.

Reply:
Got it. Expanded the javadoc in bdabd26b047ddb61a766eae93166770d53e564f8